### PR TITLE
Feature/appointments booked slots

### PR DESCRIPTION
**Pull request by aliadnanaslam7**

### CODE REVIEW

**Comment by aliadnanaslam7** [comment]
Cycle 0

File: app/controllers/appointments_controller.rb [hunk 1]
```diff
@@ -48,24 +57,34 @@ def create
...
+      end
     end
   end
 
   # PATCH /appointments/:sequence_num
   def update
     if @appointment.update(create_appointment_params.except(:doctor_id))
-      flash.now[:notice] = I18n.t('appointment.updated')
+      flash[:notice] = I18n.t('appointment.updated')
+      flag = true
```
Comment (mairamuneer39) [inline app/controllers/appointments_controller.rb]: remane variable to success

File: app/helpers/appointments_helper.rb [hunk 2]
```diff
@@ -4,6 +4,28 @@ def format_time(start_time, end_time)
...
+  end
+
+  def appointment_end_time
+    @appointment.end_time.strftime('%I:%M %p') if @appointment.end_time
+  end
+
+  def selected_date_appointments(id, date)
+    appointments = @current_hospital.appointments.pending.where(doctor_id: id)
+                                    .where('date(start_time) = ?', date.to_date)
+    return 'No appointments found' if appointments.empty?
```
Comment (mairamuneer39) [inline app/helpers/appointments_helper.rb]: move message to yml

File: app/views/appointments/_form.html.erb [hunk 3]
```diff
@@ -8,7 +8,26 @@
...
+        </thead>
+        <tbody id="current_date_appointments">
+          <% @appointments.each do |appointment| %>
+            <% if can? :read, appointment %>
+              <tr>
+                <td><%= format_time(appointment.start_time, appointment.end_time) %></td>
+              </tr>
+            <% end %>
+          <% end %>
+          <td style="border: 0"><%= 'No appointments found' if @appointments.empty? %></td>
```
Comment (mairamuneer39) [inline app/views/appointments/_form.html.erb]: same message will be used

File: app/views/documents/index.html.erb [hunk 4]
```diff
@@ -33,6 +33,7 @@
                 </tr>
               <% end %>
             <% end %>
+            <td style="border: 0"><%= 'No record found' if @documents.empty? %></td>
```
Comment (mairamuneer39) [inline app/views/documents/index.html.erb]: same

**Comment by mairamuneer39** [comment]
cycle 1

File: app/assets/javascripts/application.js [hunk 5]
```diff
@@ -24,9 +24,18 @@ $(document).on('mouseup touchend', function (e) {
   }
 });
 
+function asterisk_on_required_fields() {
```
Comment (aliadnanaslam7) [inline app/assets/javascripts/application.js]: JS camelcase

File: app/controllers/appointments_controller.rb [hunk 6]
```diff
@@ -22,13 +22,22 @@ def show
 
   # GET /appointments/new
   def new
+    @appointments = @current_hospital.appointments.pending
+                                     .where(doctor_id: params[:doctor_id])
```
Comment (aliadnanaslam7) [inline app/controllers/appointments_controller.rb]: move to model

File: app/controllers/appointments_controller.rb [hunk 7]
```diff
@@ -22,13 +22,22 @@ def show
...
+                                     .where('date(start_time) = ?',
+                                            DateTime.now.strftime('%Y-%m-%d'))
```
Comment (aliadnanaslam7) [inline app/controllers/appointments_controller.rb]: time zone

File: app/controllers/appointments_controller.rb [hunk 8]
```diff
@@ -95,16 +114,13 @@ def medical_history
...
-        format.js { render layout: false }
-      end
-      return
-    end
-    @appointment.change_status_to_cancel
-    if @appointment.save
-      flash.now[:notice] = I18n.t('appointment.cancelled')
     else
-      flash.now[:alert] = I18n.t('appointment.not_cancelled')
+      @appointment.change_status_to_cancel
```
Comment (aliadnanaslam7) [inline app/controllers/appointments_controller.rb]: method name and save

File: app/controllers/appointments_controller.rb [hunk 9]
```diff
@@ -95,16 +114,13 @@ def medical_history
...
+      if @appointment.save
+        flash.now[:notice] = I18n.t('appointment.cancelled')
```
Comment (aliadnanaslam7) [inline app/controllers/appointments_controller.rb]: success/failure

File: app/helpers/appointments_helper.rb [hunk 10]
```diff
@@ -4,6 +4,28 @@ def format_time(start_time, end_time)
...
+    @appointment.start_time.strftime('%I:%M %p') if @appointment.start_time
+  end
+
+  def appointment_end_time
+    @appointment.end_time.strftime('%I:%M %p') if @appointment.end_time
+  end
+
+  def selected_date_appointments(id, date)
+    appointments = @current_hospital.appointments.pending.where(doctor_id: id)
+                                    .where('date(start_time) = ?', date.to_date)
```
Comment (aliadnanaslam7) [inline app/helpers/appointments_helper.rb]: move to scope

File: app/models/appointment.rb [hunk 11]
```diff
@@ -68,9 +68,16 @@ def validate_appointment_schedule
...
-    return false if doctor_schedule.start_time.nil? || doctor_schedule.end_time.nil? || doctor_schedule.duration.nil?
-
-    start_time < end_time && (end_time - start_time) < doctor_schedule.duration.to_i && start_time.to_s.split[1] > doctor_schedule.start_time.to_s.split[1] && end_time.to_s.split[1] < doctor_schedule.end_time.to_s.split[1] && doctor.appointments.pending.where.not(id: id).where('(start_time BETWEEN ? AND ?) OR (end_time BETWEEN ? AND ?)', start_time, end_time, start_time, end_time).blank?
+    schedule = doctor.doctor_availabilities.find_by(day: start_time.strftime('%A'))
+    if schedule.start_time.nil? || schedule.end_time.nil? || schedule.duration.nil?
+      errors[:doctor] << I18n.t('appointment.doctor_not_available')
+      return
+    end
+
+    unless (end_time.strftime('%H%M%S%N').to_i - start_time.strftime('%H%M%S%N').to_i) < schedule.duration.utc.strftime('%H%M%S%N').to_i && start_time < end_time
```
Comment (aliadnanaslam7) [inline app/models/appointment.rb]: recheck

Comment (aliadnanaslam7) [reply app/models/appointment.rb]: Time.zone.now/datetime.now/datetime.current

File: app/views/appointments/_form.html.erb [hunk 12]
```diff
@@ -8,7 +8,26 @@
...
+        </thead>
+        <tbody id="current_date_appointments">
+          <% @appointments.each do |appointment| %>
+            <% if can? :read, appointment %>
+              <tr>
+                <td><%= format_time(appointment.start_time, appointment.end_time) %></td>
+              </tr>
+            <% end %>
+          <% end %>
+          <td style="border: 0"><%= I18n.t('app.record_not_found') if @appointments.empty? %></td>
```
Comment (aliadnanaslam7) [inline app/views/appointments/_form.html.erb]: <tr />

File: app/views/appointments/_form.html.erb [hunk 13]
```diff
@@ -8,7 +8,26 @@
...
+    <div class="col-md-4">
+      <table>
+        <thead>
+          <tr>
+            <th>Booked Slots</th>
+          </tr>
+        </thead>
+        <tbody id="current_date_appointments">
+          <% @appointments.each do |appointment| %>
+            <% if can? :read, appointment %>
```
Comment (aliadnanaslam7) [inline app/views/appointments/_form.html.erb]: recheck

File: app/views/appointments/approve.js.erb [hunk 14]
```diff
@@ -1,5 +1,6 @@
 $('#btn_approve_appointment')[0].hidden = true;
 $('#btn_cancel_appointment')[0].hidden = true;
```
Comment (aliadnanaslam7) [inline app/views/appointments/approve.js.erb]: .first / recheck

File: app/views/appointments/cancel.js.erb [hunk 15]
```diff
@@ -1,5 +1,6 @@
 $('#btn_cancel_appointment')[0].hidden = true;
```
Comment (aliadnanaslam7) [inline app/views/appointments/cancel.js.erb]: .hide()

File: app/views/appointments/show.html.erb [hunk 16]
```diff
@@ -14,12 +14,12 @@
           <% if can?(:show, @appointment.prescription) && !@appointment.prescription.nil? %>
               <%= link_to 'View Prescription', appointment_prescription_path(@appointment,@appointment.prescription), class: 'btn btn-outline-primary' %>
           <% end %>
-          <% if can?(:update, @appointment) && !@appointment.completed? %>
-            <%= link_to edit_appointment_path(@appointment, title: 'Edit'), class: 'btn btn-outline-primary', 'data-toggle': 'tooltip', title: 'Edit', id: 'btn_edit_appointment', remote: true, data: { toggle: 'modal',  target: '#modal_new_appointment' } do %>
+          <% if can?(:update, @appointment) && @appointment.requested? %>
```
Comment (aliadnanaslam7) [inline app/views/appointments/show.html.erb]: extra checks

File: app/views/shared/_alert.html.erb [hunk 17]
```diff
@@ -1 +1 @@
-<%= "<div class='alert alert-danger'>#{flash[:alert]}</div>".html_safe %>
+<%= "<div class='alert alert-danger'>#{flash[:alert][0]}</div>".html_safe %>
```
Comment (aliadnanaslam7) [inline app/views/shared/_alert.html.erb]: loop

Comment (aliadnanaslam7) [reply app/views/shared/_alert.html.erb]: discuss

Comment (aliadnanaslam7) [reply app/views/shared/_alert.html.erb]: html_safe/escape_javascript/raw/sanitize/j

File: db/schema.rb [hunk 18]
```diff
@@ -212,13 +212,16 @@
...
     t.string   "type",                   limit: 255
+    t.integer  "department_id",          limit: 4
   end
 
   add_index "users", ["confirmation_token"], name: "index_users_on_confirmation_token", unique: true, using: :btree
-  add_index "users", ["email", "hospital_id"], name: "index_users_on_email_and_hospital_id", unique: true, using: :btree
-  add_index "users", ["hospital_id"], name: "fk_rails_569ef37060", using: :btree
+  add_index "users", ["hospital_id"], name: "fk_rails_9af02237d4", using: :btree
   add_index "users", ["reset_password_token"], name: "index_users_on_reset_password_token", unique: true, using: :btree
   add_index "users", ["unlock_token"], name: "index_users_on_unlock_token", unique: true, using: :btree
```
Comment (aliadnanaslam7) [inline db/schema.rb]: add .gitignore

File: app/assets/javascripts/application.js (hunk 5 above)
Comment (aliadnanaslam7) [reply app/assets/javascripts/application.js]: done!

File: app/controllers/appointments_controller.rb (hunk 6 above)
Comment (aliadnanaslam7) [reply app/controllers/appointments_controller.rb]: done!

File: app/controllers/appointments_controller.rb (hunk 7 above)
Comment (aliadnanaslam7) [reply app/controllers/appointments_controller.rb]: done!

File: app/controllers/appointments_controller.rb (hunk 9 above)
Comment (aliadnanaslam7) [reply app/controllers/appointments_controller.rb]: done!

File: app/helpers/appointments_helper.rb (hunk 10 above)
Comment (aliadnanaslam7) [reply app/helpers/appointments_helper.rb]: done!

File: app/views/appointments/approve.js.erb (hunk 14 above)
Comment (aliadnanaslam7) [reply app/views/appointments/approve.js.erb]: done!

File: app/views/appointments/cancel.js.erb (hunk 15 above)
Comment (aliadnanaslam7) [reply app/views/appointments/cancel.js.erb]: done!

File: db/schema.rb (hunk 18 above)
Comment (aliadnanaslam7) [reply db/schema.rb]: done

File: app/views/appointments/_form.html.erb (hunk 12 above)
Comment (aliadnanaslam7) [reply app/views/appointments/_form.html.erb]: done!

File: app/controllers/appointments_controller.rb (hunk 8 above)
Comment (aliadnanaslam7) [reply app/controllers/appointments_controller.rb]: done

File: app/views/appointments/_form.html.erb (hunk 13 above)
Comment (aliadnanaslam7) [reply app/views/appointments/_form.html.erb]: done

File: app/views/shared/_alert.html.erb (hunk 17 above)
Comment (aliadnanaslam7) [reply app/views/shared/_alert.html.erb]: resolved

**Pull request merged**